### ISSUE TITLE
Make Test Error 

### ISSUE THREAD

**Issue by DrRuckus** [peridigm/peridigm#128]
Currently working on installing Peridigm-1.5 (Current Master) on a bare metal build of Ubuntu-18.04. The versions and PATH variables are as follows: 

gfortran_4.8.5
gcc_4.8.5
gcc-c++_4.8.5
Cmake_3.5.1
Openmpi_1.10.2
Mpich_3.2
Python_2.7.11
Boost_1.60.0 (I know its no longer needed)
HDF5_1.10.0
NetCDF_4.4.0
Trilinos_12.6.1

echo $PATH=

/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/bin/:/usr/local/:/usr/lib/x86_64-linux-gnu:/usr/local/bin/cmake-3.5.1/bin:/usr/local/lib/openmpi-1.10.2/bin:/usr/local/bin/hdf5-1.10.0/bin:/usr/local/bin/netcdf-4.4.0/bin:/usr/local/trilinos-12.6.1/bin

echo $LD_LIBRARY_PATH=

/usr/local/lib/openmpi-1.10.2/lib:/usr/local/lib/boost-1.60.0/lib:/usr/local/bin/hdf5-1.10.0/lib:/usr/local/bin/netcdf-4.4.0/lib:/usr/local/trilinos-12.6.1/lib

Failed tests after installing Peridigm:

57% tests passed, 80 tests failed out of 185

Total Test time (real) = 173.24 sec

The following tests FAILED:
	 12 - utPeridigm_PdQuickGridDiscretization_MPI_np2 (Failed)
	 19 - ut_QuickGrid_solidCylinder_np2 (Failed)
	 20 - ut_QuickGridMPI_np2_3x3x2 (Failed)
	 21 - ut_QuickGridMPI_np3 (Failed)
	 22 - ut_SmallMeshCylinder_np4 (Failed)
	 24 - ut_frameset_2x2x1_np4 (Failed)
	 27 - ut_Y-Z_crack_np4 (Failed)
	 28 - ut_Y-Z_crack_jacobian_np4 (Failed)
	 31 - ut_QuickGrid_loadBal_np2_3x1x1 (Failed)
	 32 - ut_QuickGrid_loadBal_np2_4x4x4 (Failed)
	 33 - ut_QuickGrid_loadBal_np8_4x4x4 (Failed)
	 35 - utReloadBalance_np4 (Failed)
	 55 - BodyForce_Implicit_np4 (Failed)
	 57 - BodyForce_Explicit_np4 (Failed)
	 59 - BodyForce_QS_np4 (Failed)
	 61 - BodyForce_QS_NOX_np4 (Failed)
	 63 - CentrifugalLoad_np4 (Failed)
	 65 - Compression_NLCGQS_3x2x2_np2 (Failed)
	 67 - Contact_Cubes_np4 (Failed)
	 69 - Contact_Cubes_Interaction_Blocks_np4 (Failed)
	 71 - Contact_Ring_np4 (Failed)
	 73 - Contact_Perforation_np3 (Failed)
	 75 - Compression_QS_3x2x2_np2 (Failed)
	 77 - Multiphysics_QS_3x2x2_np2 (Failed)
	 79 - Compression_QS_Explicit_3x2x2_np2 (Failed)
	 81 - Compression_QS_Explicit_MultiFreqOutput_3x2x2_np2 (Failed)
	 83 - Compression_QS_CyclicLoading_3x2x2_np2 (Failed)
	 85 - Compression_QS_3x2x2_Exodus_np2 (Failed)
	 87 - Compression_QS_3x2x2_TextFile_np3 (Failed)
	 89 - WaveInBar_np3 (Failed)
	 91 - WaveInBar_MultiBlock_np4 (Failed)
	 93 - Bar_OneBlock_OneMaterial_QS_np2 (Failed)
	 95 - Bar_TwoBlocks_OneMaterial_QS_np2 (Failed)
	 97 - Bar_TwoBlocks_TwoMaterial_QS_np2 (Failed)
	 99 - Bar_TwoBlocks_TwoDifferentMaterial_QS_np2 (Failed)
	100 - Bar_TwoBlocks_TwoDifferentMaterial_QS_np3 (Failed)
	102 - Bar_TwoDisconnectedPieces_QS_np2 (Failed)
	103 - Bar_TwoDisconnectedPieces_QS_np3 (Failed)
	106 - MultipleOutputFiles_np2 (Failed)
	108 - DefaultBlocks_np4 (Failed)
	110 - PrecrackedPlate_np4 (Failed)
	112 - PrecrackedPlateTwoCracks_np4 (Failed)
	114 - DiskFilter_np4 (Failed)
	116 - DeformationGradient_PlaneStrainCompression_np4 (Failed)
	118 - DeformationGradient_SimpleShear_np4 (Failed)
	120 - DeformationGradient_PureShear_np4 (Failed)
	122 - DeformationGradient_SolidRotation_np4 (Failed)
	124 - DeformationGradient_NoDeformation_np4 (Failed)
	126 - DeformationGradient_NoLoad_np4 (Failed)
	128 - Interfaces_np4 (Failed)
	131 - ShellBondFilter_np4 (Failed)
	133 - InitialDamage_np4 (Failed)
	135 - ep_cube_np2 (Failed)
	137 - ThermalExpansionCube_np4 (Failed)
	139 - ThermalExpansionBondFailure_np4 (Failed)
	141 - ThermalExpansionCubeElasticCorrespondence_np4 (Failed)
	143 - Contact_Friction_np2 (Failed)
	145 - Contact_Friction_Time_Dependent_Coefficient_np2 (Failed)
	147 - Contact_2x1x1_np2 (Failed)
	149 - BondBreakingInitialVelocity_np3 (Failed)
	151 - BondBreakingInitialVelocity-EP_np3 (Failed)
	153 - BondBreakingInitialVelocity_TimeDependentCS_np3 (Failed)
	156 - CompressionImplicit_2x1x1_np2 (Failed)
	158 - CompressionImplicitEssentialBC_2x1x1_np2 (Failed)
	160 - Compression_3x1x1_InfluenceFunction_np2 (Failed)
	161 - Compression_3x1x1_InfluenceFunction_np3 (Failed)
	163 - CompressionInitDisp_2x1x1_np2 (Failed)
	165 - VariableHorizonParsing_np4 (Failed)
	167 - MultipleHorizons_np4 (Failed)
	169 - ElasticCorrespondenceQSTension_np4 (Failed)
	170 - IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw_np1 (Failed)
	171 - IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw_np4 (Failed)
	172 - IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw_np1 (Failed)
	173 - IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw_np4 (Failed)
	175 - ViscoplasticNeedlemanFullyPrescribedTension_WithFlaw_np4 (Failed)
	177 - ViscoplasticNeedlemanFullyPrescribedTension_NoFlaw_np4 (Failed)
	179 - ElasticPlasticCorrespondenceFullyPrescribedTension_np4 (Failed)
	181 - ElasticCorrespondenceFullyPrescribedTension_np4 (Failed)
	183 - ElasticHypoelasticCorrespondenceFullyPrescribedTension_np4 (Failed)
	185 - SimpleImplicitDiffusion_np4 (Failed)
Errors while running CTest
Makefile:61: recipe for target 'test' failed
make: *** [test] Error 8

Most of the failed tests contain the np2,np3, or np4 variant of the standard np1. Where all the np1 tests pasts other than #170 and #172. 

I can promptly post any other files pertaining to this issues if required. 

Any guidance would be greatly appreciated!! 

Thank you,

Cody R. Crusenberry

**Comment by oldninja** [comment]
Hi @DrRuckus, are you using the windows subsystem for Linux (wsl) by any chance? I recall an issue with OpenMPI on wsl.

**Comment by DrRuckus** [comment]
I build a machine specifically for Peridigm using Ubuntu-18.04.

**Comment by djlittl** [comment]
A log file is generated for each test when it is run.  You should look into the log files to see exactly what the error is.

**Comment by DrRuckus** [comment]
For all the ones with np2, np3, and np4 I get an error in one form or the other as follows: 

MPI::Number of processors = 1
PdutMpiFixture teardown
Unit test runtime ERROR: ut_SmallMeshCylinder_np4 only makes sense on 4 processors
	 Re-run unit test $mpiexec -np 4 ./ut_SmallMeshCylinder_np4

**Comment by djlittl** [comment]
Please paste the error for WaveInBar_np3, which runs Peridigm itself.  The one you pasted above is a unit test.  The error handling tends to be a bit better for Peridigm itself.

**Comment by DrRuckus** [comment]
-- Peridigm
-- version 1.5.0 (Dev)


****Error on processor 0: unable to open file WaveInBar.g

/usr/local/src/peridigm-master/src/io/discretization/Peridigm_ExodusDiscretization.cpp:1136:

Throw number = 1

Throw test that evaluated to true: 1

PeridigmNS::OutputManager_ExodusII::ExodusDiscretization::loadData()() -- Error code: -1 (ex_open)
Wallclock Time (seconds):
  Total           0.0012


-- Peridigm
-- version 1.5.0 (Dev)


****Error on processor 0: unable to open file WaveInBar.g

/usr/local/src/peridigm-master/src/io/discretization/Peridigm_ExodusDiscretization.cpp:1136:

Throw number = 1

Throw test that evaluated to true: 1

PeridigmNS::OutputManager_ExodusII::ExodusDiscretization::loadData()() -- Error code: -1 (ex_open)
Wallclock Time (seconds):
  Total           0.0017


-- Peridigm
-- version 1.5.0 (Dev)


****Error on processor 0: unable to open file WaveInBar.g

/usr/local/src/peridigm-master/src/io/discretization/Peridigm_ExodusDiscretization.cpp:1136:

Throw number = 1

Throw test that evaluated to true: 1

PeridigmNS::OutputManager_ExodusII::ExodusDiscretization::loadData()() -- Error code: -1 (ex_open)
Wallclock Time (seconds):
  Total           0.0011

epu --  E Pluribus Unum
	(Out of Many One -- see http://www.greatseal.com/mottoes/unum.html)
	ExodusII Parallel Unification Program
	(Version: 4.14) Modified: 2015/11/17
Cannot open file './WaveInBar.e.3.0'
ERROR: (EPU) Problem initializing input and/or output files.

   *****************************************************************
              EXODIFF	(Version: 2.83) Modified: 2015-08-20
              Authors:  Richard Drake, rrdrake@sandia.gov           
                        Greg Sjaardema, gdsjaar@sandia.gov          
              Run on    2020/08/04   18:22:05 EDT
   *****************************************************************

ERROR: Couldn't open file "WaveInBar.e".Reading first file ... 
exodiff: ERROR: Couldn't open file "WaveInBar.e". File does not exist.

**Comment by DrRuckus** [comment]
I cannot locate the WaveinBar.e.3 file. However I do have a WaveinBar.g.3 file...

**Comment by djlittl** [comment]
It looks like Peridigm is failing right away because it is unable to open the decomposed genesis (.g) mesh files.  Because Peridigm is failing right away, it is not generating the exodus (.e) output files.

So the question is, why can’t Peridigm open the .g files?  I suggest copying the input deck (.xml or .yaml) and the decomposed mesh files (WaveInBar.g.3.0, WaveInBar.g.3.1, WaveInBar.g.3.2) into some other directory and see if you can run Peridigm directly on them.  You should double check the path to the .g files, which is listed in the input deck.  It might point to a sub directory.  Having the .g files in a sub directory might be used in the test system but isn’t common when running Peridigm directly.

**Comment by DrRuckus** [comment]
So you suggest running the WaveinBar.g files directly? Outside of the make test? 

I am new to Linux in general so it could very well be the lack of experience with the OS and the complexity of the installations itself. I would not be surprised if I have left something silly out in one of the many configuration files. 

I will run the WaveinBar.g files directly ( or at least try to ) and then post the results! 

Thank you greatly!

**Comment by djlittl** [comment]
Yes, that’s what I’m suggesting.  Basically, run the WaveInBar simulation the same way that you would use Peridigm on a simulation of your own, outside the test harness.

In serial, it’s:

Peridigm WaveInBar.xml

and in parallel it’s:

mpirun -np 3 Peridigm WaveInBar.xml

**Comment by DrRuckus** [comment]
Thank you! That helps tremendously. Again I am new to Linux and obviously Peridigm. So little command syntax like that save me a lot of time!! Once I am home I will run this and post the results. Thank you again.

**Comment by DrRuckus** [comment]
-- Peridigm
-- version 1.5.0 (Dev)

/usr/local/src/Trilinos-trilinos-release-12-6-branch/packages/teuchos/parameterlist/src/Teuchos_FileInputStream.cpp:53:

Throw number = 1

Throw test that evaluated to true: file_ == NULL

FileInputStream ctor failed to open file: WaveinBar.xml
Wallclock Time (seconds):
  Total          0.00016

**Comment by DrRuckus** [comment]
mpirun -np 3 Peridigm WaveInBar.xml

-- Peridigm
-- version 1.5.0 (Dev)


-- Peridigm
-- version 1.5.0 (Dev)


-- Peridigm
-- version 1.5.0 (Dev)

Time step (seconds):
  Stable time step    5.50136e-07
  User time step      9.0659e-08
  Safety factor       not provided 
  Time step           9.0659e-08

Total number of time steps 220

Time step (seconds):
  Stable time step    5.50136e-07
  User time step      9.0659e-08
  Safety factor       not provided 
  Time step           9.0659e-08

Total number of time steps 220

Time step (seconds):
  Stable time step    5.50136e-07
  User time step      9.0659e-08
  Safety factor       not provided 
  Time step           9.0659e-08

Total number of time steps 220

Explicit time integration [======30% Complete=======Explicit time integration [0Explicit time integration [======30% Complete=======Explicit time integration [0Explicit time inExplicit time integration [=======gration [============60% ComplExplicit time integration [===================100% Complete] 

Memory Usage (Heap Alloc MB):
  Post Execute                       12.1535
  Zoltan Search Tree                 10.9697

Wallclock Time (seconds):
  Total                                0.78
  Rebalance                         6.9e-05
  Output                               0.01
  Internal Force                       0.73
  Gather/Scatter                      0.005
  Apply Kinematic B.C.              7.8e-05
  Apply Initial Conditions          7.9e-05
  Apply Body Forces                 0.00018

Explicit time integration [===================100% Complete] 

Memory Usage (Heap Alloc MB):
  Post Execute                       12.1535
  Zoltan Search Tree                 10.9697

Wallclock Time (seconds):
  Total                                0.78
  Rebalance                         4.6e-05
  Output                               0.01
  Internal Force                       0.74
  Gather/Scatter                     0.0051
  Apply Kinematic B.C.                7e-05
  Apply Initial Conditions          7.7e-05
  Apply Body Forces                 0.00019

Explicit time integration [===================100% Complete] 

Memory Usage (Heap Alloc MB):
  Post Execute                       12.1535
  Zoltan Search Tree                 10.9697

Wallclock Time (seconds):
  Total                                0.79
  Rebalance                         7.2e-05
  Output                              0.011
  Internal Force                       0.74
  Gather/Scatter                     0.0053
  Apply Kinematic B.C.              7.8e-05
  Apply Initial Conditions          7.8e-05
  Apply Body Forces                  0.0002

**Comment by DrRuckus** [comment]
So apparently when ran with the "ctest" or the " make test" method it fails. However when I step into the /usr/local/src/Peridigm/test/regression/WaveInBar and run it as described it runs as shown in my previous comment. Regardless of how many cores I assign to the task, it runs successfully, as one would expect. 

Do I have an error within my PATH variables? Or is it something more complex?

**Comment by djlittl** [comment]
A few comments:

1) It's odd that you're seeing command-line output from all three processes.  Peridigm should be writing output to the command line for only the root process (mpi rank=0).  This isn't a huge issue, but it perhaps suggests other issues with mpi.
2) When running on N multiple ranks, Peridigm requires that the mesh file (.g) be decomposed into N parts before executing Peridigm.  This can be done with the decomp program, which is part of seacas, which you should have installed as part of Trilinos.  For this test problem, the N=1 and N=3 decompositions are already stored in the test directory, but if you try to run on any other number of mpi ranks, it should fail.  So something's not right when you say that it runs for any number of ranks (unless you're running decomp ahead of time).
3) I think you're right about PATH.  It's possible that PATH is set correctly at the command line, but is somehow not transferred to the shell environment when "make test" is run.  You could try editing one of the python scripts that runs the tests to print out the PATH to check this.  Environment variables like PATH could be slightly different from one linux distribution to another (I'm not an expert on that).

**Comment by DrRuckus** [comment]
1. Im not sure exacly what you mean by mpi rank. I assume its a synonym for the number of cores assigned? Please correct me here, this is purely an assumption. 

2. Im not certain how to test this, I can say for certain that I have setup Trilinos and the other packages exactly as instructed by the current guides found on this GitHub. 

3. I will try this next. 

All of the failed test will pass successfully if I cd directly to the .xml file and run it as you mentioned. 

Would it help if I completely format my hard drive and try again with a new install of Ubuntu?

**Comment by johntfoster** [comment]
Yes, rank = independent core/thread.

>All of the failed test will pass successfully if I cd directly to the .xml
file and run it as you mentioned.

Definitely sounds like a PATH issue.

>Would it help if I completely format my hard drive and try again with a
new install of Ubuntu?

Not likely.

**Comment by oldninja** [comment]
@DrRuckus, not sure if this helps but I followed this installation guide:
https://github.com/PeriDoX/PeriDoX/blob/master/Doc/Peridigm_Installation_Guide/Peridigm_Installation_Guide.pdf
and I just had 4 or 5 failed tests, as in #122

**Comment by johntfoster** [comment]
You're probably good then.  Those tests might need to be reexamined.

**Comment by DrRuckus** [comment]
Dr. Foster,

Thank you for your reply! 
I have formatted my hard drive and restated the entire install process 6 times now....silly me....thankfully I chose a solid state drive for this build. 

I have listed my file paths in my previous posts. I do think that djlittl is correct that its failing to generate .e files correctly. I have a guess that this is due to the PATH issue, but I am a total beginner at this so.... my hunches are probably not very well tuned yet. 

oldninja,

That is also the guide I have been using. Initially I followed it exactly but the guide is fairly dated as the wget for trilinos no longer works. Trilinos was moved to github and has to be downloaded from there now. Im also not certain what other elements are either missing or outdated from that guide. I can say 100% that following that guide exactly will result in errors. I assume this is due to constantly developing software and version compatibility.

**Comment by DrRuckus** [comment]
@oldninja I followed the posted guide explicitly, with the same exact errors. Would you please echo $PATH and echo $LD_LIBRARY_PATH and copy/paste the results of that for me? It would help me greatly. 

All of my .log files now have the errors of "undefined reference to ompi_mpi_real"

**Comment by DrRuckus** [comment]
The issue was having multiple "mpirun"s in my PATH variable. 

All but 4 tests failed. 

The following tests FAILED:
	170 - IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw_np1 (Failed)
	171 - IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw_np4 (Failed)
	172 - IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw_np1 (Failed)
	173 - IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw_np4 (Failed)

This is an issue that was inquired about on a previous thread with no answers. Any help would be appreciated!! 

Thank you all,

Cody R. Crusenberry

**Comment by DrRuckus** [comment]
Rohan,

I stopped trying to resolve these 4 errors after I was told the same.
However I have been using Peridigm for a while now and have been able to
get a few validated models out of it. I would say if all other test are
successful then consider the installation a sucess. That agrees with the
general concensus as far as I am aware. I am by no means (not even close)
to an expert however.

Thanks,

Cody R. Crusenberry

**Comment by rohandotc** [comment]
@DrRuckus Thank you for clarifying this. 

Regards,
Rohan

**Comment by jritta** [comment]
@DrRuckus @rohandotc Have a look on the issue #67.
If you compile your binary with the commit 64cbfdde0b9284e14a20af32029aeeff0bf78317 these tests do not fail, but as in the issue #67 was mentioned:
> However, potentially, the test case in verification is gonna fail. Please take a look at it.

Thus, I think only the 'gold' files are corrupt and should be regenerated with the updated material model.

@djlittl @johntfoster Should I just update the 'gold' files in these 4 test cases that we can finally get rid of the failing tests?

@mbhzdi Is the material model in the current master (cb9979dd5c83e842b130b147c44714be2dce28b3) correct now?

**Comment by mbhzdi** [comment]
Yes, the material model in the current master is correct. You can just ignore those failed tests.

**Comment by DrRuckus** [comment]
This is a very active thread at this point so I will ask. What are the odds
that anisotropic material is added to Peridgm?
